src/gui/editorwidgets/qgsqmlwidgetwrapper.cpp

```diff
@@ -36,6 +37,26 @@ bool QgsQmlWidgetWrapper::valid() const
 
 QWidget *QgsQmlWidgetWrapper::createWidget( QWidget *parent )
 {
+  QgsAttributeForm *form = qobject_cast<QgsAttributeForm *>( parent );
+
+  if ( form )
+  {
+    mFormFeature = form->feature();
+    connect( form, &QgsAttributeForm::widgetValueChanged, this, [ = ]( const QString & attribute, const QVariant & newValue, bool attributeChanged )
+    {
+      if ( attributeChanged )
+      {
+        const thread_local QRegularExpression expRe { QStringLiteral( R"re(expression.evaluate\s*\(\s*"(.*)"\))re" ), QRegularExpression::PatternOption::MultilineOption | QRegularExpression::PatternOption::DotMatchesEverythingOption };
```

Comment:
This regexp makes me nervous ...

Could we use a [`QQmlPropertyMap`](https://doc.qt.io/qt-6/qqmlpropertymap.html) instead? It's not backwards compatible but a lot cleaner.

```
QgsQmlFormObject : public QObject
{
   Q_PROPERTY( QQmlPropertyMap attributes ...)
   Q_PROPERTY( QgsGeometry geometry ... )
   Q_PROPERTY( parent and other magic ✨)
}
```

```
Label {
    text: form.attributes.name
}
```

Reply:
The regexp is the same as what's used in the HTML widget ( https://github.com/qgis/QGIS/blob/master/src/gui/editorwidgets/qgshtmlwidgetwrapper.cpp#L47 ) and the text widget follows a similar logic ( https://github.com/qgis/QGIS/blob/master/src/gui/editorwidgets/qgstextwidgetwrapper.cpp#L46 ). The logic has been in place there long enough to feel safe about it.

I love the idea of a property map, but it'd be complementary to this in any case as that wouldn't allow us to evaluate QGIS expressions which is still something that's highly desirable (ie aggregate functionality, feature fetching on any specified layer, etc).

This is fixing what we have in place. Then we can build extra features :)

Reply:
I'm sure it works in many cases. And I think it's justified for the text widget and html widgets which need a template engine to insert dynamic snippets.

The qml widget on the other hand is able to have knowledge about its content and through that offers possibilities not present for text and html. So I am hesitating if the same approach is justified. Reinitializing the qml context repeatedly during the lifetime of a qml widget could be connected to heavy processing involved in more complex scenarios.

In any case: the qml code is static, so we don't need to re-evaluate the regexp for every attribute change. I am ok with this approach if we only evaluate the regexp _once_ when the widget is initialized, so if we come up with a better integrated solution in the future it will come with only some small initial extra cost.

Reply:
@m-kuhn , ah yeah, one time evaluation, it's a good idea! It can also be applied to the other two types, it shall be done.